cnxarchive/views/xpath.py

```diff
@@ -0,0 +1,147 @@
+    resp.status = "200 OK"
+    resp.content_type = 'application/xhtml+xml'
+    resp.body = content
+    return resp
+
+
+def tree_to_html_xpath(tree):
+    """Return html list version of book tree."""
+    ul = etree.Element('ul')
+    html_listify_xpath([tree], ul)
+    return HTML_WRAPPER.format(etree.tostring(ul))
+
+
+def html_listify_xpath(tree, root_ul_element, parent_id=None):
+    """Recursively construct HTML nested list version of book tree, adding hrefs to the page titles which are links to
+    the the API endpoint which runs the xpath on that page CNXML.
+    The original caller should not call this function with the
+    `parent_id` defined.
+
+    """
+    request = get_current_request()
+    is_first_node = parent_id is None
+    if is_first_node:
+        parent_id = tree[0]['id']
+    for node in tree:
+        li_elm = etree.SubElement(root_ul_element, 'li')
+        a_elm = etree.SubElement(li_elm, 'a')
+        a_elm.text = node['title']
+        if node['id'] != 'subcol':
+            if is_first_node:
+                a_elm.set('href', request.route_path(
+                    # FIXME is there a prettier way to do this?
+                    'xpath-book', ident_hash=node['id']) + "?q=" + request.params.get('q', ''))
+            else:
+                a_elm.set('href', request.route_path(
+                    'xpath-page',
+                    ident_hash=node['id']) + "?q=" + request.params.get('q', ''))
+        if 'contents' in node:
+            elm = etree.SubElement(li_elm, 'ul')
+            html_listify_xpath(node['contents'], elm, parent_id)
+
+
+def execute_xpath(request, sql_function):
+    settings = get_current_registry().settings
+    routing_args = request and request.matchdict or {}
+
+    xpath_string = request.params.get('q', '')
+    ident_hash = routing_args.get('ident_hash', '')
+
+    if ident_hash:
+        try:
+            uuid, version = split_ident_hash(ident_hash)
+        except IdentHashShortId as e:
+            uuid = get_uuid(e.id)
+            version = e.version
+        except IdentHashMissingVersion as e:
+            uuid = e.id
+            version = get_latest_version(e.id)
+
+    results = {'results': []}
+
+    if xpath_string == '':
+        exc = httpexceptions.HTTPBadRequest()
+        exc.explanation = 'Empty xpath supplied'
+        raise exc
+
+    with psycopg2.connect(settings[config.CONNECTION_STRING]) as db_connection:
+        with db_connection.cursor() as cursor:
+
+            try:
+                cursor.execute(SQL[sql_function],
+                               {'document_uuid': uuid, 'document_version': version, 'xpath_string': xpath_string})
+            except psycopg2.Error as e:
+                exc = httpexceptions.HTTPBadRequest()
+                exc.explanation = e.pgerror
+                raise exc
+
+            for res in cursor.fetchall():
+                results['results'].append(
+                    {
+                        'module_name': res[0],
```

Comment:
I believe `cnx-archive` uses `module` and `collection` internally but I don't think the API exposes those terms. Could this just be `title:` and `uuid:`?